src/plat/pc99/machine/dev/vga/io.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2024, DornerWorks
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ */
+#include <config.h>
+#include <arch/kernel/boot_sys.h>
+#include <arch/model/statedata.h>
+#include <machine/io.h>
+#include <plat/machine/io.h>
+#include <drivers/uart.h>
+
+#define VGA_PHYS 0xb8000 // Text buffer address
+
+//black background, light grey text, no blinking
+#define VGA_COLOR 0x07
+#define CLEAR_SCREEN_CHAR 32
+static uint64_t vga_addr = 0;
+uint64_t vga_size = 0;
+// assume 80x25 console
+#define MAX_X_POS 80
+#define MAX_Y_POS 25
+
+/* Name kept 'serial_init' since other code expects to call this */
+void serial_init(uint16_t port)
+{
+    set_vga_addr(VGA_PHYS);
+}
+
+void set_vga_addr(uint64_t addr)
+{
+    vga_addr = addr;
+    // multiply by 2 since each slot in the VGA buffer has a char byte and 'style' byte
+    vga_size = MAX_X_POS * MAX_Y_POS * 2;
+}
+
+uint64_t get_vga_addr(void)
+{
+    return vga_addr;
+}
+static int verify_vga_offset(uint64_t offset)
+{
+    if(offset > vga_size)
+    {
+        printf("VGA offset of 0x%llx is outside size of 0x%llx", offset, vga_size);
```

Comment:
I think we should not print anything there, as this is called in the context of printing something? Might end in a bad recursion in the worst case.

Reply:
Good catch I didn't think about that. Should we just fail silently then?

Reply:
Could put an assert() or halt() there, as this likely a fatal configuration of implementation error.

Reply:
I put an `assert(0)`  here just to ensure it fails. I could just change it to be `assert((offset <= vga_size))` if we think that'd look cleaner

Reply:
Yes, it falls down to this in the end. But assert() also tries to print something, so do nothing seems best?

Reply:
oh right I forgot assert prints too and so does Halt. Yeah silent failing may just be the best course of action here. A user realistically shouldn't ever index outside of the allocated space if they just make calls to printf